### ISSUE TITLE
Join Node in manual stopped automatically sending joined message when last split message arrives. New bug or bug fix?

### ISSUE THREAD

**Issue by jdotbdot** [node-red/node-red#4781]
### Current Behavior

In NR3.1.9 the manual join in the flow below sends the joined string immediately after all message parts have arrived.
In NR4.0 nothing is sent unless one of the options "number of parts", "timeout" or "msg.complete" is selected.

### Expected Behavior

Perhaps this is a bugfix?
It does seem to be a breaking change and is apparently not mentioned in the documentation.

### Example flow

![image](https://github.com/node-red/node-red/assets/11013909/c48a5cdb-a760-4b4e-8775-9cfb95076db4)

```
[{"id":"262a06c7.d1a87a","type":"inject","z":"73aa4e3ce5726647","name":"Inject array","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"[\"aaa\", \"bbb\"]","payloadType":"json","x":100,"y":120,"wires":[["d1475eb351408ccd"]]},{"id":"fb44dce01fbd722e","type":"debug","z":"73aa4e3ce5726647","name":"output","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","statusVal":"","statusType":"auto","x":730,"y":120,"wires":[]},{"id":"d1475eb351408ccd","type":"split","z":"73aa4e3ce5726647","name":"default split settings","splt":"","spltType":"str","arraySplt":1,"arraySpltType":"len","stream":false,"addname":"","x":290,"y":120,"wires":[["8ca50b0b593537a2"]]},{"id":"8ca50b0b593537a2","type":"join","z":"73aa4e3ce5726647","name":"Manual join using hyphen","mode":"custom","build":"string","property":"payload","propertyType":"msg","key":"topic","joiner":"-","joinerType":"str","accumulate":false,"timeout":"","count":"","reduceRight":false,"reduceExp":"","reduceInit":"","reduceInitType":"","reduceFixup":"","x":530,"y":120,"wires":[["fb44dce01fbd722e"]]}]
```

### Environment

- Node-RED version:   3.1.9 / 4.0
- Node.js version:  20.13.1
- npm version: 10.5.2
- Platform/OS: Raspberry Pi / RPiOS
- Browser: Firefox

**Comment by dceejay** [comment]
Actually it was a "fix" introduced here - #4408 - otherwise when in manual mode it would sometimes (as-in this case) - not behave like it was under user control.
However the suggestion to maybe add an optional msg.complete to the split side could be a useful enhancement/fix for this.

**Comment by jdotbdot** [comment]
Well I deleted that suggestion from my forum post on considering that we already have msg.parts and msg.index.
Also parts might arrive at the join node out of order?

I agree it's more of a bug fix than a new bug but it is a breaking change.

**Comment by knolleary** [comment]
Can we clearly document what the change in behaviour is then? We need to do better marking prs as breaking changes as this didn't get highlighted at all in the release.

**Comment by knolleary** [comment]
Or do we need to revise the original bug fix to be an explicit option in the node rather than a change in behaviour

**Comment by dceejay** [comment]
As noted in #4408 it is a change in behaviour hence targeted at 4.x - but yes a flag/tag would be a good idea to allow breaking PRs to be more obvious at merge time, in order to get highlighted.

I think the original fix is warranted and the changed behaviour should be the default as setting something to manual should try to honour that - but that adding an option to the split node in order to add msg.complete would allow the old behaviour to be recreated more easily if required.

**Comment by Darklyte** [comment]
This also breaks `get entities` nodes since they behave similar to split nodes.

It seems like this change should give the additional option for more manual control, rather than ignoring msg.parts entirely.

**Comment by dceejay** [comment]
not sure why it should affect get entities (whatever that is ? - pointer please :-) ) - as this was a change to the join node and not the split node. The split node has not changed behaviour - the .parts property is still present after a split. The change was for the join node to ignore it.

**Comment by knolleary** [comment]
The more i look at this, I feel the *original* issue should have been addressed by adding an option to the Join node to ignore msg.parts. That would have been a much cleaner way to support that behaviour without assuming it should be the default for *everyone* who was happily using the node as-is.

**Comment by knolleary** [comment]
That said, I'm flip flopping between that and the proposal to add msg.complete on split node.

@dceejay I see you've pushed a branch is that ready to open a PR for review? Would like to ship 4.0.1 today/tomorrow - would like to have a better answer on this problem.

**Comment by jdotbdot** [comment]
I don't know the details of this msg.complete solution, but since message parts might travel different routes between the  split and join nodes,  is there any guarantee that the part with msg.complete will be the last of the set to arrive at the join node? 

For example one part might be delayed waiting for a database lookup.

**Comment by dceejay** [comment]
@knolleary - I've pushed a branch with what "I" think is a fix - but the default is still breaking from v3 in that manual mode still now means fully manual (unless msg.complete is added).  I didn't fully PR it as I'm still on vacation so haven't fully tested it... or written test cases...  does fix the simple uses cases raised when splitting an array. (and hopefully the others but...)

@jdotbdot - this issue is only an issue when the join is set in manual mode... in which case by definition you are wanting to not re-assemble things as they were before - ie you want control over when it is complete - so yes you need to work out when you want to declare it complete - and if the automatic addition of msg.complete to the last item helps or hinders you.

**Comment by jdotbdot** [comment]
The problem as I see it is if message parts arrive in the wrong order, any which missed the msg.complete cut are left hanging until the next batch of messages, which gets corrupted too.

A flow to try and emulate this:
![image](https://github.com/node-red/node-red/assets/11013909/3359124c-ff22-4b83-821f-d96e0c3aac8a)

_Every_ breakfast order that runs through this flow gets corrupted.
I know, there are other ways to convert from an array to a string.

**Comment by Darklyte** [comment]
@dceejay `get entities` is part of the home-assistant-websocket for node red (I'm not 100% sure how to reference this, but [it is here](https://github.com/zachowj/node-red-contrib-home-assistant-websocket) with [documentation here](https://zachowj.github.io/node-red-contrib-home-assistant-websocket/node/get-entities.html#split)

Essentially `get entities` pulls a bunch of data and can be set to output like a `split` node. I often use this to grab a bunch of objects from home assistant then use `join` node to concatenate them into a comma delineated string so I can command them all simultaneously.  It sends with msg.parts and that's how the `join` node knows to move forward. With manual mode completely ignoring msg.parts, there is no way for this node to function properly.

Having an option on the `join` node to ignore msg.parts when set to manual would allow for it to be used as it was previously while still giving the ability to fully control the data.

**Comment by dceejay** [comment]
@Darklyte 
Does the join node work with get entities in automatic mode ? 

But it does raise the question - my current proposed fix is to optionally add msg.complete in the join node - which would not help this as this needs the option to be to ignore msg.parts on the join node - which may be a better fix (@knolleary  ?)

**Comment by Darklyte** [comment]
@dceejay It does, in so much that it combines the payload into an array of objects. But for my use-case, at least, I'd want those entity_ids as a single string.

![Screenshot 2024-06-24 151343](https://github.com/node-red/node-red/assets/10234693/8e222760-74c1-4e90-95fd-4ba6d987204c)

**Comment by cayossarian** [comment]
> not sure why it should affect get entities (whatever that is ? - pointer please :-) ) - as this was a change to the join node and not the split node. The split node has not changed behaviour - the .parts property is still present after a split. The change was for the join node to ignore it when in manual mode.

Home Assistant Web Socket node has a widely used  [**get entities**](https://zachowj.github.io/node-red-contrib-home-assistant-websocket/node/get-entities.html#count) node that executes a split like the normal split node when a series of entities are queried, e..g., doors/windows are queried responds ... but yes the join behavior is broken that affect that pattern as well.

**Comment by dceejay** [comment]
PR #4796 should now fix this in a non-breaking way - so will be in 4.0.1 hopefully.

**Comment by knolleary** [comment]
Fixed in 4.0.1. Option added to Join node to ignore `msg.parts`. Existing flows will default to the 3.x behaviour - *not* ignoring msg.parts. New instances of the node will default to the more technically correct behaviour of ignoring msg.parts - but with an option in the node config to toggle the behaviour
